osu.Game.Rulesets.Osu/Objects/Spinner.cs

```diff
@@ -42,15 +42,10 @@ protected override void ApplyDefaultsToSelf(ControlPointInfo controlPointInfo, I
         {
             base.ApplyDefaultsToSelf(controlPointInfo, difficulty);
 
-            // spinning doesn't match 1:1 with stable, so let's fudge them easier for the time being.
-            const double stable_matching_fudge = 0.6;
-
-            // close to 477rpm
-            const double maximum_rotations_per_second = 8;
+            const double maximum_rotations_per_second = 477f / 60f;
 
             double secondsDuration = Duration / 1000;
-
-            double minimumRotationsPerSecond = stable_matching_fudge * IBeatmapDifficultyInfo.DifficultyRange(difficulty.OverallDifficulty, 3, 5, 7.5);
+            double minimumRotationsPerSecond = IBeatmapDifficultyInfo.DifficultyRange(difficulty.OverallDifficulty, 1.5, 2.5, 3.75);
```

Comment:
This is fine and all, but  the rather large elephant in the room is 2 lines below:

https://github.com/ppy/osu/blob/91c3ae4fef48dc1a3a84cb47af31f378191589df/osu.Game.Rulesets.Osu/Objects/Spinner.cs#L55

Stable counts in *half*-spins. lazer counts in *full* spins, which is reflected by the fact that the fudge is now factored into the difficulty range here. But then lazer *floors* the spin count, which means that this happens:

### Spinner 1 from [test map](https://github.com/ppy/osu/pull/24661#issuecomment-1699761763) (@ 2000ms)

```
┌────┬─────────────────────────────────┬───────────────┐
│ OD │ stable                          │ lazer         │
├────┼─────────────────────────────────┼───────────────┤
│  0 │ 12 half spins =  6   full spins │  6 full spins │
│  2 │ 15 half spins =  7.5 full spins │  7 full spins │
│  4 │ 18 half spins =  9   full spins │  9 full spins │
│  6 │ 22 half spins = 11   full spins │ 11 full spins │
│  8 │ 26 half spins = 13   full spins │ 13 full spins │
│ 10 │ 30 half spins = 15   full spins │ 15 full spins │
└────┴─────────────────────────────────┴───────────────┘
```

### Spinner 2 from [test map](https://github.com/ppy/osu/pull/24661#issuecomment-1699761763) (@ 7000ms)

```
┌────┬─────────────────────────────────┬───────────────┐
│ OD │ stable                          │ lazer         │
├────┼─────────────────────────────────┼───────────────┤
│  0 │ 27 half spins = 13.5 full spins │ 13 full spins │
│  2 │ 34 half spins = 17   full spins │ 17 full spins │
│  4 │ 41 half spins = 20.5 full spins │ 20 full spins │
│  6 │ 49 half spins = 24.5 full spins │ 24 full spins │
│  8 │ 58 half spins = 29   full spins │ 29 full spins │
│ 10 │ 67 half spins = 33.5 full spins │ 33 full spins │
└────┴─────────────────────────────────┴───────────────┘
```

tl;dr: this implementation is still potentially half a spin off (too low).

Is this fine? I don't know. But this wasn't brought up anywhere, and I figured it needs attention.

Reply:
Also for consideration is that if we change to half spins, @smoogipoo's work for anti-cheese may need changes to implementation (@smoogipoo clarification on this if possible).

Reply:
I wouldn't say I even have an implementation for anti-cheese at this point. But as for whether it's okay... probably?

Reply:
To be clear I'm not necessarily saying that we _need_ to do the half spin thing. It's _probably_ fine for lazer spinners to be half a spin easier than stable's. I just... don't know if we wanna do that and stay there, or do something else.

Reply:
Good observation. In my opinion, I think it's better to have the full spins count floored, mainly for compatibility reasons. See, with https://github.com/ppy/osu/pull/24662 in mind, flooring the full spins count would keep the bonus count matching with stable in the "odd required half spins" case, and less than stable by 1 half spin in the "even required half spins" case.

In the "odd required half spins" case, the reason why it matches stable is because in stable there's a 1.5x spin gap between completing the spinner and receiving the first bonus, and we're accommodating that by having the spinner finish 1 half spin early and add a 2x spin gap instead. Therefore the first bonus would be received at the exact moment between stable and lazer, along with all the bonus ticks afterwards.

To put the above in practice, here's a frame-by-frame shot on the first spinner of the OD0 test map, in stable:

https://github.com/ppy/osu/assets/22781491/ec4ddaee-6dec-418d-b77c-c7822c114fd8

And here it is in lazer:

https://github.com/ppy/osu/assets/22781491/cf1f34ae-e094-487a-9d6c-bde4d1ed0070

Meanwhile, in the "even required half spins" case, it's less than stable by 1 half spin due to the fact that the first bonus is gained at a half spin point, which is impossible to accommodate for in lazer while it is using full spins, so bonus gains in lazer will be shifted forward by an extra half spin compared to stable. I don't believe it's worthy of fixing in my own opinion.

Reply:
> and we're accommodating that by having the spinner finish 1 half spin early and add a 2x spin gap instead

I don't understand this part. Are you including https://github.com/ppy/osu/pull/24662 in this?

> it's less than stable by 1 half spin due to the fact that the first bonus is gained at a half spin point, which is impossible to accommodate for in lazer while it is using full spins

Well I guess part of the conversation here is whether we _want_ to go back to half spins, no?

I dunno, it is very difficult to follow that lazer is following stable that closely at all. Part of it is that stable spinners are just absolutely stupid and bugged, but the transformations you describe to arrive at the conclusion that this PR is "matching" anything are... non-trivial. Like how would I even verify what you're saying experimentally?

Reply:
> > and we're accommodating that by having the spinner finish 1 half spin early and add a 2x spin gap instead
> 
> 
> 
> I don't understand this part. Are you including https://github.com/ppy/osu/pull/24662 in this?
> 

Correct, I mentioned that first before describing each part.

Reply:
I'm reviewing this PR in isolation, because it's not even clear that we _want_ the other one yet.

Reply:
I think this is probably fine as is. We're already using full spins anyway so this PR doesn't make it any worse in that regard.

It looks like you've done some testing to make sure the values are the same, is this good to go now?

Reply:
If we're ok with the halfspin then I guess we can go. There's also the [other review thread](https://github.com/ppy/osu/pull/24661#discussion_r1310760126) that is at a bit of an impasse because it's opinion vs opinion, so either a tiebreaker would help or I can just look away.

Reply:
I'm still leaning towards fixing the half spin, but haven't had a chance to compare visuals yet.

Reply:
What does "fixing the half spin" mean here? Is it to bring half spin mechanism into osu!(lazer) and have lazer match stable calculations? Or is it to round the full spin count evenly rather than flooring it / casting to int?

Reply:
As in satisfying the change in the other PR. Making sure the spinner matches behaviour wise to stable in terms of timings.

Reply:
That should be the case already, I've explained everything in https://github.com/ppy/osu/pull/24661#discussion_r1313423259. Basically there should be nothing to change in this PR.

Reply:
I think you'd misunderstanding. I was responding directly to @bdach above. It has nothing to do with this PR but I was providing feedback on it as it may help progress this PR's change.